File: app/src/main/java/com/hwonchul/movie/presentation/view/listing/MovieAdapter.java
package com.hwonchul.movie.presentation.view.listing;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.hwonchul.movie.databinding.ItemMovieBinding;
import com.hwonchul.movie.domain.model.Movie;

import java.util.ArrayList;
import java.util.List;


public class MovieAdapter extends
        RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {
    private final List<Movie> items = new ArrayList<>();

    /**
     * 영화 상세 보기 클릭 리스너
     */
    public interface OnMovieDetailListener {
        /**
         * 상세 보기 클릭 콜백
         */
        void onClick(@NonNull Movie movie);
    }

    private OnMovieDetailListener listener;

    public void setOnMovieDetailListener(OnMovieDetailListener listener) {
        this.listener = listener;
    }

    public void setItems(@NonNull List<Movie> newItems) {
        items.clear();
        items.addAll(newItems);
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemMovieBinding binding = ItemMovieBinding.inflate(
                LayoutInflater.from(parent.getContext()), parent, false);
        return new MovieViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        final Movie movie = items.get(position);
        holder.bind(movie);
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public class MovieViewHolder extends RecyclerView.ViewHolder {
        private final ItemMovieBinding binding;

        public MovieViewHolder(@NonNull ItemMovieBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(Movie movie) {
            binding.setItem(movie);
            binding.setRank(getBindingAdapterPosition());

            binding.layoutPoster.setOnClickListener(__ -> listener.onClick(movie));
        }
    }
}

File: app/src/main/java/com/hwonchul/movie/presentation/listing/MovieAdapter.kt
package com.hwonchul.movie.presentation.listing

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.hwonchul.movie.databinding.ItemMovieBinding
import com.hwonchul.movie.domain.model.Movie
import com.hwonchul.movie.presentation.listing.MovieAdapter.MovieViewHolder

class MovieAdapter : RecyclerView.Adapter<MovieViewHolder>() {

    private val items: MutableList<Movie> = ArrayList()

    interface OnMovieDetailListener {
        fun onClick(movie: Movie)
    }

    private lateinit var listener: OnMovieDetailListener

    fun setOnMovieDetailListener(listener: OnMovieDetailListener) {
        this.listener = listener
    }

    fun setItems(newItems: List<Movie>) {
        items.clear()
        items.addAll(newItems)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
        val binding = ItemMovieBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return MovieViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
        val movie = items[position]
        holder.bind(movie)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    inner class MovieViewHolder(
        private val binding: ItemMovieBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(movie: Movie) {
            binding.item = movie
            binding.rank = bindingAdapterPosition
            binding.layoutPoster.setOnClickListener { listener.onClick(movie) }
        }
    }
}